src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -21,9 +23,38 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceProvider
 
         public async Task Invoke(HttpContext httpContext)
         {
-            using (var container = RequestServicesContainer.EnsureRequestServices(httpContext, _services))
+            // All done if we already have a request services
+            if (httpContext.RequestServices != null)
             {
-                await _next.Invoke(httpContext);
+                return;
+            }
+
+            var serviceProvider = httpContext.ApplicationServices ?? _services;
+
+            var appServiceProvider = serviceProvider.GetRequiredService<IServiceProvider>();
+            var appServiceScopeFactory = serviceProvider.GetRequiredService<IServiceScopeFactory>();
+
+            if (serviceProvider != appServiceProvider)
+            {
+                appServiceProvider = serviceProvider;
+                appServiceScopeFactory = serviceProvider.GetRequiredService<IServiceScopeFactory>();
+            }
+
+            try
+            {
+                // Creates the scope and tempororarily swap services
+                using (var scope = appServiceScopeFactory.CreateScope())
+                {
+                    httpContext.ApplicationServices = appServiceProvider;
+                    httpContext.RequestServices = scope.ServiceProvider;
+
+                    await _next.Invoke(httpContext);
+                }
+            }
+            finally
+            {
+                httpContext.RequestServices = serviceProvider; // REVIEW: should this go back to null instead?
```

Comment:
The old request services value is always null since we no-op otherwise.  I think this was copied from code that was meant to be general purpose, and we no longer are general purpose, so I think this should just be null?